File: laundry/wash_wallet/packages/wash_wallet_domain/lib/src/entities/customer_subscription.dart
import 'package:equatable/equatable.dart';
import 'customer.dart';
import 'service_package.dart';
import 'customer_quota.dart';

class CustomerSubscription extends Equatable {
  final int id;
  final int customerId;
  final int servicePackageId;
  final String subscriptionCode;
  final double pricePaid;
  final DateTime? purchaseDate;
  final DateTime? expiredAt;
  final String status;
  final DateTime? createdAt;
  final DateTime? updatedAt;
  final int? remainingDays;
  final bool isUnlimited;
  final String statusBadgeVariant;
  final String statusLabel;
  final Customer? customer;
  final ServicePackage? servicePackage;
  final List<CustomerQuota> customerQuotas;

  const CustomerSubscription({
    required this.id,
    required this.customerId,
    required this.servicePackageId,
    required this.subscriptionCode,
    required this.pricePaid,
    this.purchaseDate,
    this.expiredAt,
    required this.status,
    this.createdAt,
    this.updatedAt,
    this.remainingDays,
    this.isUnlimited = false,
    this.statusBadgeVariant = 'secondary',
    this.statusLabel = '',
    this.customer,
    this.servicePackage,
    this.customerQuotas = const [],
  });

  factory CustomerSubscription.fromModel(dynamic model) {
    return CustomerSubscription(
      id: model.id,
      customerId: model.customerId,
      servicePackageId: model.servicePackageId,
      subscriptionCode: model.subscriptionCode,
      pricePaid: model.pricePaid,
      purchaseDate: model.purchaseDate,
      expiredAt: model.expiredAt,
      status: model.status,
      createdAt: model.createdAt,
      updatedAt: model.updatedAt,
      remainingDays: model.remainingDays,
      isUnlimited: model.isUnlimited,
      statusBadgeVariant: model.statusBadgeVariant,
      statusLabel: model.statusLabel,
      customer: model.customer != null
          ? Customer.fromModel(model.customer)
          : null,
      servicePackage: model.servicePackage != null
          ? ServicePackage.fromModel(model.servicePackage)
          : null,
      customerQuotas:
          (model.customerQuotas as List?)
              ?.map((e) => CustomerQuota.fromModel(e))
              .toList() ??
          [],
    );
  }

  @override
  List<Object?> get props => [
    id,
    customerId,
    servicePackageId,
    subscriptionCode,
    pricePaid,
    purchaseDate,
    expiredAt,
    status,
    createdAt,
    updatedAt,
    remainingDays,
    isUnlimited,
    statusBadgeVariant,
    statusLabel,
    customer,
    servicePackage,
    customerQuotas,
  ];
}

File: laundry/wash_wallet_be/resources/js/types/customer_subscription.ts
import {
    BaseFilters,
    BaseSortOptions,
    Customer,
    CustomerQuota,
    QuotaUsageLog,
    ServicePackage,
} from ".";

export type SubscriptionStatus =
    | "active"
    | "exhausted"
    | "expired"
    | "canceled";

export type SubscriptionStatusLabel =
    | "Aktif"
    | "Habis"
    | "Kadaluarsa"
    | "Dibatalkan";

export type StatusBadgeVariant = "success" | "warning" | "danger" | "secondary";

export interface CustomerSubscription {
    id: number;
    customerId: number;
    servicePackageId: number;
    subscriptionCode: string;
    pricePaid: number;
    purchaseDate: string;
    expiredAt: string | null;
    purchaseDateFrom?: string;
    purchaseDateTo?: string;
    status: SubscriptionStatus;
    statusLabel?: SubscriptionStatusLabel;
    statusBadgeVariant: StatusBadgeVariant;
    isUnlimited: boolean;
    createdAt: string;
    updatedAt: string;
    customer?: Customer;
    servicePackage?: ServicePackage;
    customerQuotas: CustomerQuota[];
    quotaUsageLogs?: QuotaUsageLog[];
    remainingDays?: number | null;
    isExpired?: boolean;
    isExhausted?: boolean;
}

/**
 * Customer subscription filter interface
 */
export interface CustomerSubscriptionFilters extends BaseFilters {
    customerId?: number;
    servicePackageId?: number;
    status?: SubscriptionStatus;
    subscriptionCode?: string;
    purchaseDate?: string;
    expiredAt?: string;
    purchaseDateFrom?: string;
    purchaseDateTo?: string;
    expiringInDays?: number;
    hasRemainingQuota?: boolean;
    withRelations?: boolean;
    withCounts?: boolean;
    expiredAtFrom?: string;
    expiredAtTo?: string;
}

/**
 * Customer subscription sort options
 */
export interface CustomerSubscriptionSortOptions extends BaseSortOptions {
    column:
        | "subscriptionCode"
        | "pricePaid"
        | "purchaseDate"
        | "expiredAt"
        | "status"
        | "createdAt"
        | "updatedAt"
        | "customerId"
        | "servicePackageId";
}

/**
 * Customer subscription form data
 */
export interface CustomerSubscriptionFormData {
    customerId: number;
    servicePackageId: number;
    pricePaid: number;
    purchaseDate: string;
    expiredAt?: string | null;
}
